// ngboost-auth/esm2022/lib/auth.service.mjs
import { throwError } from 'rxjs';
import { catchError, tap } from 'rxjs/operators';
import { Injectable, Inject } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "./cookie.service";
import * as i3 from "@angular/router";
export class AuthService {
    constructor(http, cookieService, r, auth_urls) {
        this.http = http;
        this.cookieService = cookieService;
        this.r = r;
        this.auth_urls = auth_urls;
        this.cookieOpts = {
            // domain: 'localhost',
            path: '/',
            expires: 3,
            // expires: new Date('2018-10-31T03:24:00'),
            secure: false,
            httpOnly: false,
            sameSite: 'strict'
        };
    }
    /**
     * Login with username and password
     * @param creds // credentials object {username: xxx, password: xxx}
     * @return Observable
     */
    login(creds) {
        return this.http.post(this.auth_urls.apiLoginUrl, creds).pipe(tap((apiResp) => {
            /* set cookie 'auth_jwtToken': 'JWT xyz...' */
            if (!!apiResp && !!apiResp.jwtToken) {
                this.cookieService.put('auth_jwtToken', apiResp.jwtToken, this.cookieOpts, false);
            }
            /* set cookie 'auth_loggedUser' and class property 'this.loggedUser': {first_name: , last_name: , ...} */
            if (!!apiResp && !!apiResp.loggedUser) {
                this.cookieService.putObject('auth_loggedUser', apiResp.loggedUser, this.cookieOpts, false);
                this.loggedUser = apiResp.loggedUser;
            }
            /* redirect to URL */
            const afterGoodLoginURL = this.auth_urls.afterGoodLogin.replace('{loggedUserRole}', apiResp.loggedUser.role);
            this.r.navigateByUrl(afterGoodLoginURL);
        }), catchError((err) => {
            /* remove all cookies */
            this.cookieService.removeAll(false);
            /* remove loggedUser class property */
            this.loggedUser = null;
            // return error
            return throwError(() => err);
        }));
    }
    /**
     * Logout
     * @return Observable
     */
    logout() {
        // delete all cookies
        this.cookieService.removeAll(false);
        // delete class property
        this.loggedUser = null;
        // redirect
        setTimeout(() => {
            const afterLogoutURL = this.auth_urls.afterLogout;
            this.r.navigateByUrl(afterLogoutURL);
        }, 300);
    }
    /**
     * Get logged user info (from global variables or cookie)
     * @return object - {first_name: , last_name: , ...}
     */
    getLoggedUserInfo() {
        const loggedUser = this.loggedUser || this.cookieService.getObject('auth_loggedUser', false);
        return loggedUser;
    }
    /**
     * Get JWT token from cookie
     * @return string - JWT eyJhbGciOiJIUzI1NiIsInR...
     */
    getJWTtoken() {
        const jwtToken = this.cookieService.get('auth_jwtToken', false) || 'JWT ';
        return jwtToken;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.3", ngImport: i0, type: AuthService, deps: [{ token: i1.HttpClient }, { token: i2.CookieService }, { token: i3.Router }, { token: 'AUTH_URLS' }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.3", ngImport: i0, type: AuthService }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.3", ngImport: i0, type: AuthService, decorators: [{
            type: Injectable
        }], ctorParameters: () => [{ type: i1.HttpClient }, { type: i2.CookieService }, { type: i3.Router }, { type: undefined, decorators: [{
                    type: Inject,
                    args: ['AUTH_URLS']
                }] }] });
