// ngboost-auth/esm2022/lib/jwtTokenInterceptor.service.mjs
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./auth.service";
/**
 * Intercept every HTTP erquest.
 */
export class JwtTokenInterceptor {
    constructor(authService) {
        this.authService = authService;
    }
    intercept(req, next) {
        const jwtToken = this.authService.getJWTtoken();
        req = req.clone({
            setHeaders: {
                Authorization: jwtToken
            }
        });
        return next.handle(req);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.3", ngImport: i0, type: JwtTokenInterceptor, deps: [{ token: i1.AuthService }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.3", ngImport: i0, type: JwtTokenInterceptor }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.3", ngImport: i0, type: JwtTokenInterceptor, decorators: [{
            type: Injectable
        }], ctorParameters: () => [{ type: i1.AuthService }] });
