// ngboost-auth/esm2022/lib/ngboost-auth.module.mjs
import { NgModule } from '@angular/core';
import { AuthService } from './auth.service';
import { CookieService } from './cookie.service';
import { JwtTokenInterceptor } from './jwtTokenInterceptor.service';
import { IsLoggedService } from './routeGuards/isLogged.service';
import { HasRoleService } from './routeGuards/hasRole.service';
import { AutologinService } from './routeGuards/autoLogin.service';
import * as i0 from "@angular/core";
export class NgboostAuthModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.3", ngImport: i0, type: NgboostAuthModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.3", ngImport: i0, type: NgboostAuthModule }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.3", ngImport: i0, type: NgboostAuthModule, providers: [
            AuthService,
            CookieService,
            JwtTokenInterceptor,
            IsLoggedService,
            HasRoleService,
            AutologinService,
        ] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.3", ngImport: i0, type: NgboostAuthModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [],
                    declarations: [],
                    providers: [
                        AuthService,
                        CookieService,
                        JwtTokenInterceptor,
                        IsLoggedService,
                        HasRoleService,
                        AutologinService,
                    ],
                    exports: []
                }]
        }] });
