// ngboost-auth/esm2022/lib/routeGuards/autoLogin.service.mjs
import { Injectable, Inject } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../auth.service";
import * as i2 from "@angular/router";
/**
 * Check if user is logged and if yes do auto login e.g. redirect to afterGoodLogin URL.
 * This guard service apply only on /login route.
 */
export class AutologinService {
    constructor(authService, r, auth_urls) {
        this.authService = authService;
        this.r = r;
        this.auth_urls = auth_urls;
    }
    canActivate(ars, rss) {
        // get loggedUser info after successful username:password login
        const loggedUser = this.authService.getLoggedUserInfo();
        // console.log('GUARD::autoLogin:loggedUser: ', loggedUser);
        const isAlreadyLogged = !!loggedUser && !!loggedUser.username;
        // redirect
        if (isAlreadyLogged) {
            const afterGoodLoginURL = this.auth_urls.afterGoodLogin.replace('{loggedUserRole}', loggedUser.role);
            this.r.navigateByUrl(afterGoodLoginURL);
        }
        return true; // always activate this route
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.3", ngImport: i0, type: AutologinService, deps: [{ token: i1.AuthService }, { token: i2.Router }, { token: 'AUTH_URLS' }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.3", ngImport: i0, type: AutologinService }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.3", ngImport: i0, type: AutologinService, decorators: [{
            type: Injectable
        }], ctorParameters: () => [{ type: i1.AuthService }, { type: i2.Router }, { type: undefined, decorators: [{
                    type: Inject,
                    args: ['AUTH_URLS']
                }] }] });
