// ngboost-auth/esm2022/lib/routeGuards/hasRole.service.mjs
import { Injectable, Inject } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../auth.service";
import * as i2 from "@angular/router";
/**
 * Check if user has required role: admin, customer. If not redirect to /login page.
 */
export class HasRoleService {
    constructor(authService, r, auth_urls) {
        this.authService = authService;
        this.r = r;
        this.auth_urls = auth_urls;
    }
    canActivate(ars, rss) {
        // get loggedUser info after successful username:password login
        const loggedUser = this.authService.getLoggedUserInfo();
        // console.log('GUARD::hasRole:loggedUser: ', loggedUser);
        // get current URL and check if user's role (admin, customer) is contained in it
        const currentUrl = rss.url; // '/admin/dashboard'
        // console.log('GUARD::hasRole:currentUrl: ', currentUrl);
        let tf = false;
        if (loggedUser) {
            tf = currentUrl.indexOf(loggedUser.role) !== -1;
        }
        if (!tf) {
            this.r.navigateByUrl(this.auth_urls.afterBadLogin);
            console.error(new Error('Guard blocks this route because user doesn\'t have good role. Redirection to /login page.'));
        }
        return tf;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.3", ngImport: i0, type: HasRoleService, deps: [{ token: i1.AuthService }, { token: i2.Router }, { token: 'AUTH_URLS' }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.3", ngImport: i0, type: HasRoleService }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.3", ngImport: i0, type: HasRoleService, decorators: [{
            type: Injectable
        }], ctorParameters: () => [{ type: i1.AuthService }, { type: i2.Router }, { type: undefined, decorators: [{
                    type: Inject,
                    args: ['AUTH_URLS']
                }] }] });
